### PR TITLE
Improve REST error handling when endpoint does not support HTTP verb, add OPTIONS support

### PR DESCRIPTION
**Pull request by dakrone**
This is a follow-up of #17916 from @jbertouch's work brought up to date with the latest master, here's a subset of original description:

-----
First pass at closing #15335. The new behavior for the `RestController#executeHandler` method is as follows:
- For a request to a **valid endpoint** with an **unsupported HTTP method** -> Return a _405 HTTP error_ including allowed methods for the endpoint in the **Allow** HTTP header. Refer to [HTTP/1.1 - 10.4.6 - 405 Method Not Allowed](https://tools.ietf.org/html/rfc2616#section-10.4.6).
- For an OPTIONS HTTP method request to a **valid endpoint** -> Return a _200 HTTP response_ including allowed methods for the endpoint in the **Allow** HTTP header. Refer to [HTTP/1.1 - 9.2 - Options](https://tools.ietf.org/html/rfc2616#section-9.2).
----

It allows things like the following:

```
~ λ curl -v -XPOST 'localhost:9200/my_index/_settings'
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 9200 (#0)
> POST /my_index/_settings HTTP/1.1
> Host: localhost:9200
> User-Agent: curl/7.51.0
> Accept: */*
> 
< HTTP/1.1 405 Method Not Allowed
< Allow: PUT,GET
< content-type: application/json; charset=UTF-8
< content-length: 134
< 
{
  "error" : "Incorrect HTTP method for uri [/my_index/_settings] and method [POST], allowed: [PUT, GET]",
  "status" : 405
}
* Curl_http_done: called premature == 0
* Connection #0 to host localhost left intact
```

Where you can see the `HTTP/1.1 405 Method Not Allowed` response and `Allow: GET` header.

It also adds support for `OPTIONS`:

```
~ λ curl -v -XOPTIONS x:9200/_cluster/settings
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to x (127.0.0.1) port 9200 (#0)
> OPTIONS /_cluster/settings HTTP/1.1
> Host: x:9200
> User-Agent: curl/7.51.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Allow: GET,PUT
< content-type: text/plain; charset=UTF-8
< content-length: 0
< 
* Curl_http_done: called premature == 0
* Connection #0 to host x left intact
```

### CODE REVIEW

**Comment by dakrone** [comment]
So, I think there's more that can be done overall to improve REST error handling, in particular, handling the NPE generated when ES interprets the request as an index request with an empty body, but I think that can be done in a separate PR rather than making this larger.

**Comment by jasontedor** [comment]
> So, I think there's more that can be done overall to improve REST error handling, in particular, handling the NPE generated when ES interprets the request as an index request with an empty body, but I think that can be done in a separate PR rather than making this larger.

There's already a PR open for this (#23497) but I want to make sure that included in the effort here is not incorrectly routing requests to the index endpoint? See my previous [comment](https://github.com/elastic/elasticsearch/pull/17916#issuecomment-242996804) on this topic.

**Comment by jbertouch** [comment]
Hi @jasontedor, thank you for following up on this. In reference to your previous comment on incorrect index request routing:

> requests to POST /{index}/_settings are still being routed to an index action, matching the POST handler for /{index}/{type} but ideally these would 405 since there is a PUT handler for {index}/_settings.

I added  4 `PathTrie.TrieMatchingMode` modes to handle the routing error you identified. Requests are matched against the the most restrictive mode first (ie. only explicit nodes are matched), and step through to the least restrictive (ie. wildcard nodes are matched anywhere).

In the case you identified, the intermediate `TrieMatchingMode#WILDCARD_ROOT_NODES_ALLOWED` mode should pick up the POST to `/{index}/_settings` as intended for the handler registered for PUT requests, and return a 405. The handler attached to `/{index}/{type}` only matches for the  `TrieMatchingMode#WILDCARD_NODES_ALLOWED` mode, so it does not match the request.

I added a test for this scenario at `RestHttpResponseHeadersIT#testIndexSettingsPostRequest`, which was passing in #17916, but I haven't run it against this PR yet.

**Comment by dakrone** [comment]
> I added a test for this scenario at RestHttpResponseHeadersIT#testIndexSettingsPostRequest, which was passing in #17916, but I haven't run it against this PR yet.

When I originally merged master there were conflicts and this was passing, but with the latest master merge it seems two tests in `RestHttpResponseHeadersIT` are failing, I am having a look to fix them.

**Comment by dakrone** [comment]
I discovered an issue in this PR in the course of fixing the tests today, the original PR also suffers from this:

When searching for a working endpoint, if an endpoint matches without the matching method, we can reject the request with a `405` because the verb didn't match, even though an empty param could have caused a match, for example with these two registered endpoints (that have different handlers):

```
POST - "/_alias"
GET - "/_alias/{name}"
```

With a request `GET /_alias`, this assumes the endpoint is the `POST` variant and issues a 405 response, when in reality it should have treated the `GET` part as valid with an empty `name` parameter.

I'm looking into fixing this.

**Comment by jbertouch** [comment]
Hi @dakrone, in the example you mentioned, if `GET /_alias` should match the handler at `GET /_alias/{name}`, is there a reason why the same handler isn't explicitly registered to `GET /_alias` too? Curious because path variables usually default to being required for request mapping in web frameworks (eg. Spring), so I'm used to seeing that pattern.

That question aside, as leaf wildcard nodes are currently always matched if present ([PathTrie.java#L155](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/path/PathTrie.java#L155)), could the solution be as simple as removing or changing the priority of the `TrieMatchingMode#EXPLICIT_NODES_ONLY` strategy?

**Comment by dakrone** [comment]
> is there a reason why the same handler isn't explicitly registered to GET
> /_alias too?

I don't believe there's any reason other than our historic code convention,
certainly no reason it couldn't be explored in this PR if it makes the matching
explicit.

> That question aside, ... could the solution be as simple as removing or
> changing the priority of the TrieMatchingMode#EXPLICIT_NODES_ONLY strategy?

I did quickly try changing the priority of the Trie matching attempts, however
it caused errors in other REST tests due to matching the wrong pattern.

I do plan to work on this though, so don't take it as me giving up :)

**Comment by dakrone** [comment]
>> is there a reason why the same handler isn't explicitly registered to GET
>> /_alias too?

> I don't believe there's any reason other than our historic code convention,
> certainly no reason it couldn't be explored in this PR if it makes the
> matching explicit.

I experimented with this, for most of the failures, this seemed to work,
however, there was one more issue with this. Consider this scenario:

- `RestPutAliasesAction` registers `PUT /_alias`, as expected.

- `RestGetAliasesAction` is changed to register `GET /_alias` so it correctly
  handles the leaf node version

- `RestGetIndicesAction` registers `GET /{index}`, however, it's slightly more
  complex than that...

`RestGetIndicesAction` actually parses the `{index}` above to check for the
presence of things starting with `_`, such as "_alias", "_settings", or
"_alias,_mappings", which it then uses in dispatch for writing particular parts
of the response.

The issue is that we basically have a duplication here, between
`RestGetAliasesAction` without an index name, and `RestGetIndicesAction` with
`_alias` as the "feature" to return. Since `RestGetIndicesAction` doesn't
technically register a `GET /_alias` endpoint, it goes to the
`RestGetAliasesAction` instead (with the new registration), and therefore the
behavior is completely different.

Personally, I think this is not a good way for `RestGetIndicesAction` to
register endpoints, rather than register a wildcard and then try to parse the
different parts from a string like "_settings,_alias,_mappings".

This, however, would be a breaking change, so I'm curious about others' opinions
for how to proceed in this case.

**Comment by jasontedor** [comment]
> This, however, would be a breaking change, so I'm curious about others' opinions
for how to proceed in this case.

+1, we don't need the complexity on this endpoint. The endpoint should send everything back and users can filter what they don't need.

**Comment by dakrone** [comment]
@javanna this is finally ready for review!

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 1]
```diff
@@ -233,31 +215,66 @@ public void dispatchBadRequest(
...
+                    handleUnsupportedHttpMethod(request, channel, validMethodSet);
+                    return true;
+                } else if (!validMethodSet.contains(request.method())
+                        && request.method() == RestRequest.Method.OPTIONS) {
+                    handleOptionsRequest(request, channel, validMethodSet);
+                    return true;
                 } else {
-                    final String msg = "No handler found for uri [" + request.uri() + "] and method [" + request.method() + "]";
-                    channel.sendResponse(new BytesRestResponse(BAD_REQUEST, msg));
+                    return false;
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: or maybe give the return value a variable and name it accordingly

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 2]
```diff
@@ -233,31 +215,66 @@ public void dispatchBadRequest(
...
             if (handler == null) {
-                if (request.method() == RestRequest.Method.OPTIONS) {
-                    // when we have OPTIONS request, simply send OK by default (with the Access Control Origin header which gets automatically added)
-
-                    channel.sendResponse(new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY));
+                /*
+                 * Get the map of matching handlers for a request, for the full set of HTTP methods.
+                 */
+                // TODO: don't explicitly use this TrieMatching method
+                final HashSet<RestRequest.Method> validMethodSet = getValidHandlerMethodSet(request, trieMatchingMode);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: looking at the `getValidHandlerMethodSet` and `getHandlersForMethod` methods, I wonder if we have the right datastructure for this in place. I wonder if we should have a single `PathTrie<MethodHandlers> handler` trie that holds a single object as the _output_ of the trie that allows you to look up all the handlers for the same path at once instead of going back an trying all methods one after another. We can use the methods ordinal to lookup stuff in an array and just leave it null if we don't have one. 
If we do that we can do this first as a separate change?

File: core/src/main/java/org/elasticsearch/common/path/PathTrie.java [hunk 3]
```diff
@@ -26,6 +26,28 @@
 import static java.util.Collections.unmodifiableMap;
 
 public class PathTrie<T> {
+    
+    public enum TrieMatchingMode {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: it would we nice if we can make this an impl detail and just offer a method that returns a list or a stream of `<T>` instead of exposing this explicitly.

**Review by s1monw — CHANGES_REQUESTED** [review]
I left some commetns and questions

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 4]
```diff
@@ -308,11 +325,115 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
...
+        final Map<String, String> originalParams = new HashMap<>(request.params());
+
+        // Request execution flag
+        boolean requestHandled = false;
+
+        /*
+         * First try handlers mapped to explicit paths only.
+         */
+        if (requestHandled == false) {
+            final RestHandler handler = getHandler(request, TrieMatchingMode.EXPLICIT_NODES_ONLY);
```
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I experimented with doing this for quite a while the other day, unfortunately, it turned out to be too difficult to do cleanly with the way that `PathTrie` currently behaves with `RestHandler`. Right now attempting to retrieve a handler from the trie actually modifies the `RestRequest`, which means that trying to return a list of handlers is very side-effect-y and even if we select the right handler, there request may be out of sync because it was modified so many times.

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 2 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: > If we do that we can do this first as a separate change?

I'm unclear about which one you mean to do first and which one as a separate change, can you clarify which order you mean?

File: core/src/main/java/org/elasticsearch/common/path/PathTrie.java (hunk 3 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: This falls under the same request-mutating issue as the previous comment

**Comment by dakrone** [comment]
Thanks for taking a look @s1monw, I pushed a commit and left some comments in response to your feedback, let me know what you think!

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 4 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I do think that as a subsequent change we should try to make PathTrie and RestHandler a little more decoupled when it comes to mutation and param parsing, I think it would be too complex for this (already complex) PR though

**Comment by s1monw** [comment]
@dakrone  you can re-request reviews then I'd see this earlier

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I think we should fix our datastrucuture first and don't make Path trie super complicated and flexible. This should be fixed first before we make this change here.

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 4 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I think this should happen first to make this PR less complex

**Comment by dakrone** [comment]
@s1monw okay! I've pushed two commits that make some drastic changes to this PR:

The first is a change from `PathTrie<RestHandler>` to `PathTrie<MethodHandlers>` so that we can have a single `PathTrie` for all paths, instead of a separate one for each method. It allows us to retrieve the supported methods for a path in a much easier manner (this was from your suggestion).

The second is a change in `PathTrie` (and of course `RestController`) to make it retrieve *all* of the `RestHandlers` (now `MethodHandlers` actually) at once, looping through the candidates and trying each to see if it is valid. This is also from your suggestion. As a byproduct, no outside classes have to know anything about `TrieMatchingMode` (again, your suggestion) and the class has been made package-private.

File: core/src/main/java/org/elasticsearch/common/path/PathTrie.java [hunk 5]
```diff
@@ -220,10 +349,59 @@ public T retrieve(String path, Map<String, String> params) {
...
+        private final List<TrieMatchingMode> modes = new ArrayList<>(4);
+        private final Supplier<Map<String, String>> paramSupplier;
+        private final PathTrie<T> trie;
+        private final String path;
+
+        PathTrieIterator(PathTrie trie, String path, Supplier<Map<String, String>> paramSupplier) {
+            this.path = path;
+            this.trie = trie;
+            this.paramSupplier = paramSupplier;
+            modes.add(TrieMatchingMode.EXPLICIT_NODES_ONLY);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: just use Arrays.asList when this is initializaed?

File: core/src/main/java/org/elasticsearch/common/path/PathTrie.java [hunk 6]
```diff
@@ -220,10 +349,59 @@ public T retrieve(String path, Map<String, String> params) {
...
+            modes.add(TrieMatchingMode.WILDCARD_ROOT_NODES_ALLOWED);
+            modes.add(TrieMatchingMode.WILDCARD_LEAF_NODES_ALLOWED);
+            modes.add(TrieMatchingMode.WILDCARD_NODES_ALLOWED);
+            assert TrieMatchingMode.values().length == 4 : "missing trie matching mode";
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (modes.isEmpty() == false) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: `return modex.isEmpty() == false`?

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 7]
```diff
@@ -141,31 +142,39 @@ public void registerWithDeprecatedHandler(RestRequest.Method method, String path
...
+    }
+
+    /**
+     * Registers a REST handler to be executed when the provided method and path match the request.
+     *
+     * @param singleMethod GET, POST, etc.
+     * @param path Path to handle (e.g., "/{index}/{type}/_bulk")
+     * @param handler The handler to actually execute
+     */
+    public void registerHandler(RestRequest.Method singleMethod, String path, RestHandler handler) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: why did you add this method and why don't you use the `registerHandler` method above instead?

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 8]
```diff
@@ -308,32 +326,110 @@ boolean checkRequestParameters(final RestRequest request, final RestChannel chan
...
+        boolean requestHandled = false;
+
+        if (checkRequestParameters(request, channel) == false) {
+            channel.sendResponse(BytesRestResponse.createSimpleErrorResponse(channel,
+                            BAD_REQUEST, "error traces in responses are disabled."));
+            requestHandled = true;
+        }
+
+        // Loop through all possible handlers, attempting to dispatch the request
+        Iterator<MethodHandlers> allHandlers = getAllHandlers(request);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: hmm can't we do `for (MethodHandlers handler : ((Iterable<MethodHandlers>) this::getAllHandlers))`?

**Review by s1monw — CHANGES_REQUESTED** [review]
I took another look and I like this. Yet, I think we need to make the change that you did here first in a separate PR and then do the actual change based on that. Is this possible? I'd love to make is smaller so reviews are more effective.

File: core/src/main/java/org/elasticsearch/common/path/PathTrie.java (hunk 5 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: The modes need to go in a specific order, this ensures that even if someone changes the order in the ENUM definition the matching will still work correctly

Comment (dakrone) [inline core/src/main/java/org/elasticsearch/common/path/PathTrie.java]: Oh, but I see what you're saying, you're just saying for contructing it, sure I'll change that

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 8 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I can't cast the Iterator to Iterable (I tried it and got class cast exceptions), is there a different way to get an Iterable out of an Iterator?

File: core/src/main/java/org/elasticsearch/rest/RestController.java (hunk 7 above)
Comment (dakrone) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: This is a compatibility layer for all the `Rest*` actions, if I used only the one above I would have to change every action's `registerHandler` method call (which I might do, but in a subsequent PR)

**Comment by dakrone** [comment]
> Yet, I think we need to make the change that you did here first in a separate PR and then do the actual change based on that.

I opened #25459 as a spin-off encapsulating the `MethodHandlers` and `PathTrie` changes

**Comment by dakrone** [comment]
@s1monw #25459 was merged and I've merged master into this branch, bringing it from +915 lines to +369 (much smaller!), please take another look!

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 9]
```diff
@@ -246,10 +246,19 @@ boolean dispatchRequest(final RestRequest request, final RestChannel channel, fi
...
             }
         } else {
-            if (request.method() == RestRequest.Method.OPTIONS) {
-                // when we have OPTIONS request, simply send OK by default (with the Access Control Origin header which gets automatically added)
-
-                channel.sendResponse(new BytesRestResponse(OK, BytesRestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY));
+            // Get the map of matching handlers for a request, for the full set of HTTP methods.
+            final Set<RestRequest.Method> validMethodSet = getValidHandlerMethodSet(request);
+            if (validMethodSet.size() > 0
+                && !validMethodSet.contains(request.method())
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: I know its controversial but can we use `== false` ?

File: core/src/main/java/org/elasticsearch/rest/RestController.java [hunk 10]
```diff
@@ -246,10 +246,19 @@ boolean dispatchRequest(final RestRequest request, final RestChannel channel, fi
...
+                && request.method() != RestRequest.Method.OPTIONS) {
+                // If an alternative handler for an explicit path is registered to a
+                // different HTTP method than the one supplied - return a 405 Method
+                // Not Allowed error.
+                handleUnsupportedHttpMethod(request, channel, validMethodSet);
+                requestHandled = true;
+            } else if (!validMethodSet.contains(request.method())
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/rest/RestController.java]: same here ^^

**Review by s1monw — APPROVED** [review]
this looks good to me. Yet, I think we should have a real REST test for this somehow? WDYT @dakrone we can add a test REST handler in a smoketest qa env where we can add a rest spec that we don't implement, something along those lines? This can be a folllowup IMO

**Comment by dakrone** [comment]
> Yet, I think we should have a real REST test for this somehow? WDYT @dakrone we can add a test REST handler in a smoketest qa env where we can add a rest spec that we don't implement, something along those lines? 

Sure I will open an issue to add this as a follow-up

Thanks for the reviews @s1monw!

**Comment by dakrone** [comment]
> Yet, I think we should have a real REST test for this somehow? WDYT @dakrone we can add a test REST handler in a smoketest qa env where we can add a rest spec that we don't implement, something along those lines? 

Definitely, I opened #25611 for this.

Thanks for the reviews @s1monw!
